### PR TITLE
Dispatcher logger is primary and BaseYii logger is secondary

### PR DESCRIPTION
**Pull request by antonmarin**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13637, #12163

### CODE REVIEW

**Comment by antonmarin** [comment]
Seems like pr is ready now :)

**Comment by cebe** [comment]
This is a breaking change, so if we fix it, this should be done in 2.1.

**Comment by antonmarin** [comment]
@cebe it breaks only if you try logging before application initialization. Is it possible?

**Comment by samdark** [comment]
Yes. It is what's done automatically for `trace` level if `YII_DEBUG` is true. It logs all the app init sequence.

**Comment by antonmarin** [comment]
@samdark isn't log components bootstapped at this time?

**Comment by samdark** [comment]
It should be.

**Comment by antonmarin** [comment]
@samdark so, do i understand right that it will be ok if log component should always bootstrap because it used while application initialization? What about i add log component to BaseApplication->bootstrap?

**Comment by samdark** [comment]
I think so. @cebe, @SilverFire, @dynasource, @klimov-paul what do you think?

**Comment by dynasource** [comment]
IMO, the Application should not bootstrap anything by default at all. It should be a slim as possible, ideally just enough to survive.

Regarding the logging. This should be refactored in the framework to be optional, like is requested is several topics throughout the issue tracker. The problem is performance and hickups during processing of large data. It is hard for people to understand why it is occurring and to eventually understand it to be happing because of a Logger (which they seemed to have disabled).

**Comment by antonmarin** [comment]
@dynasource as @samdark said there is logging all the app init sequence, so is it possible to make logger optional? I think the only way is to create some DummyLogger or remove logging from app init. If you prefer first way, so my PR is actual, and the other way is to remove both logging and bootstrap logger. Have i missed something?

**Comment by antonmarin** [comment]
any updates?

**Comment by samdark** [comment]
Actually, I've re-checked code and I'm not so sure if we're actually logging init sequence prior to log is initialized. A good way to check it is to set logging level to trace, enable debug, run a single request and save logs. Then check if logs are the same with this pull request.

**Comment by antonmarin** [comment]
`yii\base\Controller:130 Yii::trace('Route to run: ' . $action->getUniqueId(), __METHOD__);`

**Comment by antonmarin** [comment]
well what direction to move? set logger optional and check existence (not so pretty)? create dummy? or leave log bootstrapping?

**Comment by antonmarin** [comment]
`\yii\base\Application:322 Yii::trace('Bootstrap with ' . get_class($component), __METHOD__);`

**Comment by antonmarin** [comment]
Another way is to require Logger config as basePath

**Comment by samdark** [comment]
Leave log bootstrapping. Bootstrap sequence is important and needs logging.

**Comment by antonmarin** [comment]
ok. so what about merge? :)

File: tests/framework/BaseYiiTest.php [hunk 1]
```diff
@@ -99,8 +100,7 @@ public function testSetupLogger()
         $this->assertSame($logger, BaseYii::getLogger());
 
         BaseYii::setLogger(null);
-        $defaultLogger = BaseYii::getLogger();
-        $this->assertTrue($defaultLogger instanceof Logger);
+        BaseYii::getLogger();
```
Comment (SilverFire) [inline tests/framework/BaseYiiTest.php]: `\yii\base\InvalidConfigException` may be thrown on `Logger` initialization as well. This test should be split

File: framework/BaseYii.php [hunk 2]
```diff
@@ -362,7 +364,7 @@ public static function getLogger()
             return self::$_logger;
         }
 
-        return self::$_logger = static::createObject('yii\log\Logger');
+        throw new InvalidConfigException('No logger without application');
```
Comment (SilverFire) [inline framework/BaseYii.php]: I think this exception will be popular so we should explicitly explain what's going on.

How about `Logger component must be initialized before using logging actions. Add "log" component to the "bootstrap" section of the configuration on the first position.`?

**Review by SilverFire — CHANGES_REQUESTED** [review]

**Review by dynasource — CHANGES_REQUESTED** [review]
I would expect to see unit tests that reflect the title & the changes in this PR

**Comment by antonmarin** [comment]
@dynasource, i don't understand. Could you explain in different words?

**Comment by dynasource** [comment]
the title states "Same logger in Yii and Yii::$app->log->logger" but I can not see this by looking at the new unit tests. I can only see new tests about throwing errors which does not go into depth about your actual usecase.

**Comment by antonmarin** [comment]
> DispatcherTest:62 already reflects current changes
$this->assertSame(Yii::getLogger(), $dispatcher->getLogger());

So, do i understang right, that you mean, that that test should be moved to separate test method?

**Comment by samdark** [comment]
I think he means that the test should assert that logger is the same in both cases.

**Comment by dynasource** [comment]
correct, and otherwise the title of this issue should be changed towards the real goal of this PR

**Comment by antonmarin** [comment]
> I think he means that the test should assert that logger is the same in both cases.
> $this->assertSame(Yii::getLogger(), $dispatcher->getLogger());
isn't that do it?

**Comment by antonmarin** [comment]
so i change title as test already exists

**Comment by antonmarin** [comment]
any update?

File: framework/log/Dispatcher.php [hunk 3]
```diff
@@ -113,7 +113,7 @@ public function init()
     public function getLogger()
     {
         if ($this->_logger === null) {
-            $this->setLogger(Yii::getLogger());
+            $this->setLogger('yii\log\Logger');
```
Comment (samdark) [inline framework/log/Dispatcher.php]: `\yii\log\Logger::class` could be used.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/log/Dispatcher.php (hunk 3 above)
Comment (antonmarin) [inline framework/log/Dispatcher.php]: But Yii requires php 5.4?

Comment (sergeymakinen) [inline framework/log/Dispatcher.php]: This PR is aimed for Yii 2.1 which requires PHP 5.5 ATM.

Comment (TerraSkye) [inline framework/log/Dispatcher.php]: 2.1 is set for 7.1 if im not mistaken https://github.com/yiisoft/yii2/issues/11397

Comment (cebe) [inline framework/log/Dispatcher.php]: @TerraSkye correct, but that decision has been made after @sergeymakinen commented ;)

**Comment by samdark** [comment]
I'm thinking about ditching our logger altogether and using PSR-3 compatible interface with monolog as default implementation.

**Comment by sergeymakinen** [comment]
@samdark I couldn't agree more. Even though I love the Yii logger.

**Comment by dynasource** [comment]
Yeah. Good idea

**Comment by antonmarin** [comment]
great idea, so maybe i should close PR?

**Comment by samdark** [comment]
Yes. Would be cool if you'll create another issue about using PSR-3 and reference this one. Thanks.

**Comment by samdark** [comment]
Won't be merged into 2.0.

**Pull request closed without merging**